src/irmin-pack/unix/chunked_suffix.ml

```diff
@@ -0,0 +1,189 @@
+      with OpenInventoryError err ->
+        Error (err : open_error :> [> open_error ])
+  end
+
+  type t = { inventory : Inventory.t }
+
+  (* A lightweight wrapper around creating/opening append only files as chunks *)
+  module Ao_chunk = struct
+    let path = Layout.V4.suffix_chunk
+
+    let create_rw ~root ~chunk_idx ~overwrite ~auto_flush_threshold
+        ~auto_flush_procedure =
+      let path = path ~root ~chunk_idx in
+      Ao.create_rw ~path ~overwrite ~auto_flush_threshold ~auto_flush_procedure
+
+    let open_rw ~root ~chunk_idx ~auto_flush_threshold ~auto_flush_procedure
+        ~end_poff ~dead_header_size =
+      let path = path ~root ~chunk_idx in
+      Ao.open_rw ~path ~end_poff ~dead_header_size ~auto_flush_threshold
+        ~auto_flush_procedure
+
+    let open_ro ~root ~chunk_idx ~end_poff ~dead_header_size ~is_appendable =
+      let open Result_syntax in
+      let path = path ~root ~chunk_idx in
+      (* Passing [end_poff] is required by Ao for its consistency check (and
+         only applies to the appendable chunk).
+         TODO: We could change its API in the future so we can not do
+         [Io.size_of_path], which is admittedly very silly.*)
+      let* end_poff =
+        if is_appendable then Ok end_poff else Io.size_of_path path
+      in
+      Ao.open_ro ~path ~end_poff ~dead_header_size
+  end
+
+  let create_rw ~root ~start_idx ~overwrite ~auto_flush_threshold
+      ~auto_flush_procedure =
+    let open Result_syntax in
+    let chunk_idx = start_idx in
+    let+ ao =
+      Ao_chunk.create_rw ~root ~chunk_idx ~overwrite ~auto_flush_threshold
+        ~auto_flush_procedure
+    in
+    let chunk = { idx = chunk_idx; suffix_off = Int63.zero; ao } in
+    let inventory = Inventory.v 1 (Fun.const chunk) in
+    { inventory }
+
+  let open_rw ~root ~end_poff ~start_idx ~chunk_num ~dead_header_size
+      ~auto_flush_threshold ~auto_flush_procedure =
+    let open Result_syntax in
+    let open_chunk ~chunk_idx ~is_appendable =
+      match is_appendable with
+      | true ->
+          Ao_chunk.open_rw ~root ~chunk_idx ~end_poff ~auto_flush_threshold
+            ~auto_flush_procedure ~dead_header_size
+      | false ->
+          Ao_chunk.open_ro ~root ~chunk_idx ~end_poff ~dead_header_size
+            ~is_appendable
+    in
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let open_ro ~root ~end_poff ~dead_header_size ~start_idx ~chunk_num =
+    let open Result_syntax in
+    let open_chunk = Ao_chunk.open_ro ~root ~end_poff ~dead_header_size in
+    let+ inventory = Inventory.open_ ~start_idx ~chunk_num ~open_chunk in
+    { inventory }
+
+  let appendable_ao t = (Inventory.appendable t.inventory).ao
+  let end_poff t = appendable_ao t |> Ao.end_poff
+
+  let read_exn t ~off ~len buf =
+    let chunk = Inventory.find ~off t.inventory in
+    let poff = suffix_off_to_chunk_poff chunk off in
+    Ao.read_exn chunk.ao ~off:poff ~len buf
```

Comment:
Instead of an assert, I [changed](https://github.com/mirage/irmin/compare/43929d1b048439db9c65ce4ab820757b3e4fdfd0..8f2229f96581fc3dcd742869770c868268bd5825) `find` to return a `poff` and moving the conversion code into that function as well. If the reader is unconvinced, they can look in the `find` code, which should give everything they need.